txpool/account.go

```diff
@@ -293,3 +295,33 @@ func (a *account) promote() []*types.Transaction {
 
 	return promoted
 }
+
+// resetSkips sets 0 to skips
+func (a *account) resetSkips()  {
+	a.skips = 0
+}
+
+// incrementSkips increments skips
+func (a *account) incrementSkips()  {
+	a.skips++
+}
+
+// getLowestTx returns the transaction with lowest nonce, which might be popped next
+// this method don't pop a transaction from both queues
+func (a *account) getLowestTx() *types.Transaction {
+	a.promoted.lock(true)
+	defer a.promoted.unlock()
+
+	if firstPromoted := a.promoted.peek(); firstPromoted != nil {
+		return firstPromoted
+	}
+
+	a.enqueued.lock(true)
+	defer a.enqueued.unlock()
+
+	if firstEnqueued := a.enqueued.peek(); firstEnqueued != nil {
+		return firstEnqueued
+	}
+
```

Comment:
This can potentially brick an account later on when `Drop` sets the nonce to that of the enqueued transaction. 

Here is my proposal on `updateAccountSkipsCounts` and `getLowestTx`:

```
	p.accounts.Range(
		func(key, value interface{}) bool {
			address, _ := key.(types.Address)
			account, _ := value.(*account)

			if _, ok := latestActiveAccounts[address]; ok {
				account.resetSkips()

				return true
			}

			firstPromoted := account.getLowestTx()
			if firstPromoted == nil {
				// no need to increment anything,
				// account has no txs
				return true
			}

			account.incrementSkips()

			if account.skips < maxAccountSkips {
				return true
			}

			// account has been skipped too many times
			p.Drop(firstPromoted)
			account.resetSkips()

			return true
		},
	)
```

```
func (a *account) getLowestTx() *types.Transaction {
	a.promoted.lock(true)
	defer a.promoted.unlock()

	return a.promoted.peek()
}
```

With these two, only accounts which are **skipped and contain promoted transactions** are processed. Skipped account with no transactions do not fit the criteria, so no need to increment `skips`.

Reply:
Fixed 5386a85520ccf47e44e84fbdb6f4ff9e3f4cd407